curated_transformers/models/hf_hub/conversion.py

```diff
@@ -1,15 +1,278 @@
+        out = dict(hf_config)
+        out.update(dataclasses.asdict(self))
+        return out
+
+
+@dataclass
+class HFConfigKey:
+    """
+    Metadata for a Hugging Face configuration key that describes
+    how it must be (de-)serialized.
+
+    :param name:
+        Name of the key in the Hugging Face config file.
+    :param mapping_to_curated_config:
+        Either a string or a tuple of a string and a callable.
+        The string is the name of the keyword argument passed
+        to the constructor of the Curated Transformers config and
+        the callable accepts the value of the key and returns a
+        converted value.
+    :param mapping_from_curated_config:
+        A callable that accepts the config instance and returns
+        a converted value.
+    """
+
+    name: str
+    mapping_to_curated_config: Union[str, Tuple[str, Callable[[Any], Any]]]
+    mapping_from_curated_config: Callable[[Any], Any]
+
+    @property
+    def curated_config_kwarg(self) -> str:
+        """
+        Returns the name of the Curated Transformer config keyword
+        argument.
+        """
+        return (
+            self.mapping_to_curated_config
+            if isinstance(self.mapping_to_curated_config, str)
+            else self.mapping_to_curated_config[0]
+        )
+
+    def get_kwarg(self, kwargs: Dict[str, Any]) -> Any:
+        """
+        Looks up the value of the key in the given keyword
+        argument dictionary.
+
+        :param kwargs:
+            Keyword argument dictionary.
+        :returns:
+            Value of the key.
+        """
+        return kwargs[self.curated_config_kwarg]
+
+    def set_kwarg(self, value: Union["HFConfigKey", Any], kwargs: Dict[str, Any]):
+        """
+        Converts and sets the value of the keyword argument in
+        the given dictionary.
+
+        :param value:
+            Either a config key instance or a value. In the case
+            of the former, its value is looked up in the keyword
+            argument dictionary and saved as the calling config
+            key's value as-is. In the case of the latter, the
+            value is first converted (if possible) before being
+            set.
+        :param kwargs:
+            Keyword argument dictionary.
+        """
+        if isinstance(value, HFConfigKey):
+            kwargs[self.curated_config_kwarg] = value.get_kwarg(kwargs)
+            return
+
+        if isinstance(self.mapping_to_curated_config, tuple):
+            curated, ctor = self.mapping_to_curated_config
+        else:
+            curated = self.mapping_to_curated_config
+            ctor = lambda x: x
+
+        kwargs[curated] = ctor(value)
+
+    def remove_kwarg(self, kwargs: Dict[str, Any]):
+        """
+        Removes the key the given keyword argument dictionary.
+
+        :param kwargs:
+            Keyword argument dictionary.
+        """
+        kwargs.pop(self.curated_config_kwarg)
+
+
+@dataclass
+class HFConfigKeyDefault:
```

Comment:
Why a class for this?

Reply:
The presence of a default value indicates whether a particular HF config key is optional. We could simply use an `Optional[Any]` type to encode the default value generically, but some keys might use `None` itself as a default value. So, we need a wrapper type to disambiguate such cases.